### PR TITLE
Add a generator to the PVeRA reparameterization function.

### PR DESCRIPTION
**Pull request by leofillioux**
This PR addresses [this comment](https://github.com/huggingface/peft/pull/3189#discussion_r3140569649) from PR #3189.

It integrates a generator in the reparameterization function from PVeRA, and uses it for sampling the random epsilon. This generator (`torch.Generator`) can be specified in the config.

### CODE REVIEW

File: src/peft/tuners/pvera/config.py [hunk 1]
```diff
@@ -187,6 +190,7 @@ class PveraConfig(PeftConfig):
             ),
         },
     )
+    generator: torch.Generator = field(default=None, metadata={"help": "Random generator for sampling from the learned distribution."})
```
Comment (BenjaminBossan) [inline src/peft/tuners/pvera/config.py]: This wouldn't allow to save the config as a JSON, right? How about having a `seed` argument instead that, if it's not `None`, serves as a seed for the generator?

**Review by BenjaminBossan — CHANGES_REQUESTED** [review]
Thanks for the quick PR. I think there is a small issue, otherwise it looks good. Please check.

**Comment by leofillioux** [comment]
Hey @BenjaminBossan, thanks for your review! I agree, I hadn't thought of that. I switched to generator_seed, and removed unnecessary arguments in the _reparameterize function. What do you think?

File: src/peft/tuners/pvera/layer.py [hunk 2]
```diff
@@ -146,7 +152,8 @@ def reset_pvera_parameters(self, adapter_name, d_initial: float = 0.1):
     def _reparametrize(self, mu, logvar, sample_at_inference):
         if self.training or (not self.training and sample_at_inference):
             std = torch.exp(0.5 * logvar)
-            eps = torch.randn_like(std)
+            print(self.generator)
```
Comment (BenjaminBossan) [inline src/peft/tuners/pvera/layer.py]: :see_no_evil:

Comment (leofillioux) [reply src/peft/tuners/pvera/layer.py]: Oh wow sorry about that haha

**Comment by leofillioux** [comment]
@BenjaminBossan I removed that debugging line 😊

**Comment by BenjaminBossan** [comment]
@leofillioux please run `make style`.

**Review by BenjaminBossan — CHANGES_REQUESTED** [review]
Just noticed that it would be good to test this. Let's just add a single pvera test to the end of `TestPeftCustomModel` (i.e. [here](https://github.com/huggingface/peft/blob/9a20e07d347fa5f1c565f3393aa0c4fea469badb/tests/test_custom_models.py#L3794)) and check for reproducibility.

**Comment by leofillioux** [comment]
@BenjaminBossan any idea how to check for reproducibility? Because going twice through the model with a generator won't lead to the same results

**Comment by BenjaminBossan** [comment]
> going twice through the model with a generator won't lead to the same results

Oh, why is that?

**Comment by leofillioux** [comment]
The state of the generator changes every time I go through it, so if I go twice through the model, the result won't be the same.

I thought about making two instances of the model, except that the model needs to be trained for one iteration for the adaptation to be non zero. Maybe I should just create a deep copy of the model after the update.

**Comment by leofillioux** [comment]
@BenjaminBossan here is my implementation for the test.

File: tests/test_custom_models.py [hunk 3]
```diff
@@ -3792,6 +3792,25 @@ def test_mha_gradients_set_correctly(self, with_forward_call):
         assert model.base_model.model.mha.base_layer.out_proj.base_layer.weight.requires_grad is True
         assert model.base_model.model.mha.base_layer.in_proj_weight.requires_grad is True
 
+    def test_pvera_reproducible(self):
+        inputs = {"input_ids": torch.arange(10).view(-1, 1).to(self.torch_device)}
+        config = PveraConfig(r=4, sample_at_inference=True, generator_seed=0)
```
Comment (BenjaminBossan) [inline tests/test_custom_models.py]: I get the idea of training the model so that it's not a no-op, but passing `init_weights=False` should be enough, right? The way I would test is more like:

```python
config = PveraConfig(init_weights=False, generator_seed=0, ...)
model = AutoModelForCausalLM.from_pretrained(...)
model = get_peft_model(model, config)
output0 = model(inputs)

del model

config = PveraConfig(init_weights=False, generator_seed=0, ...)
model = AutoModelForCausalLM.from_pretrained(...)
model = get_peft_model(model, config)
output1 = model(inputs)

# check that output0 and output1 are identical
```

I think this better reflects what users of PVeRA would expect when they re-run an experiment and want reproducibility.

Comment (leofillioux) [reply tests/test_custom_models.py]: The problem when using `init_weight=False` is that `pvera_lambda_d` is randomly initialised to a different vector for both models. I could set the seed before instantiating the model, but then the predictions would be equal regardless of setting `generator_seed` to None or a specified integer.

Comment (BenjaminBossan) [reply tests/test_custom_models.py]: I see. I think it would be valid to set the same seed before `get_peft_model`, as that is expected to achieve reproducibility. Then you can set the seed to two different values before calling `foward`, in which case we should still see the same result when the generator is being used, right?

**Comment by leofillioux** [comment]
@BenjaminBossan, thanks for the feedback! Here is an updated version of the test.

**Comment by BenjaminBossan** [comment]
Thanks, the test looks good @leofillioux. Could you please merge with/rebase on the latest main branch? The CI should finally be green again :)

**Review by BenjaminBossan — APPROVED** [review]
Thanks for adding this feature quickly, LGTM.
